hw/xbox/nv2a_shaders.c

```diff
@@ -23,50 +23,166 @@
 #include "hw/xbox/nv2a_shaders_common.h"
 #include "hw/xbox/nv2a_shaders.h"
 
-static void generate_geometry_shader_pass_vertex(QString* s, const char* v)
+static QString* generate_geometry_shader(
+                                      enum ShaderPolygonMode polygon_front_mode,
+                                      enum ShaderPolygonMode polygon_back_mode,
+                                      enum ShaderPrimitiveMode primitive_mode,
+                                      GLenum *gl_primitive_mode)
 {
-    qstring_append_fmt(s, "        gl_Position = gl_in[%s].gl_Position;\n", v);
-    qstring_append_fmt(s, "        gl_PointSize = gl_in[%s].gl_PointSize;\n", v);
-    qstring_append_fmt(s, "        g_vtx = v_vtx[%s];\n", v);
-    qstring_append(s,     "        EmitVertex();\n");
-}
 
-static QString* generate_geometry_shader(enum ShaderPrimitiveMode primitive_mode)
-{
-    /* generate a geometry shader to support deprecated primitive types */
-    QString* s = qstring_new();
-    qstring_append(s, "#version 330\n");
-    qstring_append(s, "\n");
-    switch (primitive_mode) {
-    case PRIM_TYPE_QUADS:
-        qstring_append(s, "layout(lines_adjacency) in;\n");
-        qstring_append(s, "layout(triangle_strip, max_vertices = 4) out;\n");
-        break;
-    default:
-        assert(false);
-        break;
+    /* FIXME: Missing support for 2-sided-poly mode */
+    assert(polygon_front_mode == polygon_back_mode);
+    enum ShaderPolygonMode polygon_mode = polygon_front_mode;
+
+    /* POINT mode shouldn't require any special work */
+    if (polygon_mode == POLY_MODE_POINT) {
+        *gl_primitive_mode = GL_POINTS;
+        return NULL;
     }
-    qstring_append(s, "\n");
-    qstring_append(s, STRUCT_VERTEX_DATA);
-    qstring_append(s,
-        "noperspective in VertexData v_vtx[];\n");
-    qstring_append(s,
-        "noperspective out VertexData g_vtx;\n");
-    qstring_append(s, "\n");
 
-    qstring_append(s, "void main() {\n");
+    /* Handle LINE and FILL mode */
+    const char *layout_in;
+    const char *layout_out;
+    const char *body;
```

Comment:
I thought about it but I'd rather get a warning if they are set and not used
or
warning if they are not set but used
[which the compiler should do for the current code]

Whenever `layout_in`, `layout_out` or `body` are used the code should reach the geometry shader generator. If they are not set you should `return NULL` before.
However, I'll probably init to NULL to make sure that the assertions before the shader gen. will work [which effectivly turns compiler warnings into assert-cases]